File: JavaScript/Ransom Note.js
// https://leetcode.com/problems/ransom-note/

/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 *
 * canConstruct("aa", "aab") -> true
 */
var canConstruct = function(ransomNote, magazine) {
  var 
    i = 0, 
    len = magazine.length, 
    cache = {}
  
  while (i < len) {
    if (!cache[magazine[i]]) {
      cache[magazine[i]] = 0
    }
    cache[magazine[i]]++
    i++
  }

  i = 0
  len = ransomNote.length
  
  while (i < len) {
    if (!cache[ransomNote[i]]) {
      return false
    }
    cache[ransomNote[i]]--
    if (cache[ransomNote[i]] < 0) {
      return false
    }
    i++
  }
 
  return true
};

// console.log(canConstruct('', 'a'))

File: 383-ransom-note/ransom_note.go
package solution

func canConstruct(ransomNote string, magazine string) bool {
	m := make(map[rune]int)
	for _, b := range magazine {
		m[b]++
	}
	for _, b := range ransomNote {
		m[b]--
		if m[b] < 0 {
			return false
		}
	}
	return true
}
